config/v2/validation.go

```diff
@@ -41,229 +41,80 @@ func (c *Config) Validate() error {
+			errs = multierror.Append(errs, fmt.Errorf("provider profile %s ", component))
 		}
-	}
-	return s
-}
 
-// lastNonNilStringSlice, despite its name can return nil if all results are nil
-func lastNonNilStringSlice(getter func(Common) []string, commons ...Common) []string {
-	var s []string
-	for _, c := range commons {
-		t := getter(c)
-		if t != nil {
-			s = t
+		if p.Version == nil {
+			errs = multierror.Append(errs, fmt.Errorf("provider version for %s ", component))
 		}
-	}
-	return s
-}
-
-// ResolveRequiredString will resolve the value and panic if it is nil. Only to be used after validations are run.
-func ResolveRequiredString(getter func(Common) *string, commons ...Common) string {
-	return *lastNonNil(getter, commons...)
-}
-
-// ResolveRequiredInt will resolve the value and panic if it is nil. Only to be used after validations are run.
-func ResolveRequiredInt64(getter func(Common) *int64, commons ...Common) int64 {
-	return *lastNonNilInt64(getter, commons...)
-}
-
-func ResolveOptionalString(getter func(Common) *string, commons ...Common) *string {
-	return lastNonNil(getter, commons...)
-}
-
-func ResolveOptionalStringSlice(getter func(Common) []string, commons ...Common) []string {
-	return lastNonNilStringSlice(getter, commons...)
-}
 
-func ResolveStringArray(def []string, override []string) []string {
-	if override != nil {
-		return override
-	}
-	return def
-}
-
-func ResolveStringMap(getter func(Common) map[string]string, commons ...Common) map[string]string {
-	resolved := map[string]string{}
-
-	for _, c := range commons {
-		m := getter(c)
-		for k, v := range m {
-			resolved[k] = v
+		if p.AccountID == nil || *p.AccountID == 0 {
+			errs = multierror.Append(errs, fmt.Errorf("provider account id for %s", component))
 		}
+		return errs
 	}
-	return resolved
-}
-
-// validateInheritedStringField will walk all accounts and components and ensure that a given field is valid at at least
-// one level of the inheritance hierarchy. We should eventually distinuish between not present and invalid because
-// if the value is present but invalid we should probably mark it as such, rather than papering over it.
-func (c *Config) validateInheritedStringField(fieldName string, getter func(Common) *string, validator func(string) bool) *multierror.Error {
-	var err *multierror.Error
 
 	// For each account, we need the field to be valid in either the defaults or account
-	for acctName, acct := range c.Accounts {
-		v := lastNonNil(getter, c.Defaults.Common, acct.Common)
+	for name, acct := range c.Accounts {
+		v := ResolveAWSProvider(c.Defaults.Common, acct.Common)
 		if v == nil {
-			err = multierror.Append(err, fmt.Errorf("account %s must have field %s at either the account or defaults level", acctName, fieldName))
-		} else if !validator(*v) {
-			err = multierror.Append(err, fmt.Errorf("account %s must have a valid %s set at either the account or defaults level", acctName, fieldName))
+			// nothing to validate
```

Comment:
I feel like these branches should live inside the validate function

Reply:
working on it in the next PR